File: topics/graphs/medium/0207-course-schedule/solution.cpp
// 0207 · Course Schedule
// Approach: DFS cycle detection | Time: O(V+E) | Space: O(V+E)

#include <vector>
#include <queue>
using namespace std;

class Solution {
    vector<vector<int>> graph;
    vector<int> state;

    bool dfs(int node) {
        if (state[node] == 1) return false;
        if (state[node] == 2) return true;
        state[node] = 1;
        for (int neighbor : graph[node])
            if (!dfs(neighbor)) return false;
        state[node] = 2;
        return true;
    }
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        graph.assign(numCourses, {});
        state.assign(numCourses, 0);
        for (auto& p : prerequisites)
            graph[p[1]].push_back(p[0]);
        for (int i = 0; i < numCourses; i++)
            if (!dfs(i)) return false;
        return true;
    }
};

File: topics/graphs/medium/0207-course-schedule/solution.py
# 0207 · Course Schedule
# Approach: DFS cycle detection | Time: O(V+E) | Space: O(V+E)

from typing import List
from collections import defaultdict, deque


class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = defaultdict(list)
        for a, b in prerequisites:
            graph[b].append(a)

        # 0 = unvisited, 1 = in path, 2 = done
        state = [0] * numCourses

        def dfs(node):
            if state[node] == 1:
                return False
            if state[node] == 2:
                return True
            state[node] = 1
            for neighbor in graph[node]:
                if not dfs(neighbor):
                    return False
            state[node] = 2
            return True

        return all(dfs(i) for i in range(numCourses))


# Approach 2: BFS / Kahn's algorithm | Time: O(V+E) | Space: O(V+E)
class SolutionBFS:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = defaultdict(list)
        indegree = [0] * numCourses
        for a, b in prerequisites:
            graph[b].append(a)
            indegree[a] += 1

        queue = deque(i for i in range(numCourses) if indegree[i] == 0)
        processed = 0
        while queue:
            node = queue.popleft()
            processed += 1
            for neighbor in graph[node]:
                indegree[neighbor] -= 1
                if indegree[neighbor] == 0:
                    queue.append(neighbor)
        return processed == numCourses
